### PR TITLE
CoreLogMessage - structured logging from Rust

### LINKED ISSUE #233: Forward Rust logs to Swift
Requirements:
https://github.com/Co-Epi/app-ios/issues/232 
https://github.com/Co-Epi/app-backend-rust/issues/27
https://github.com/Co-Epi/app-backend-rust/issues/28

The idea is that the app registers a callback in rust (probably as part of `NativeCore.bootstrap`) and we call our global `log` with these logs.

Ideally Rust should level + text and forward in the app with the respective log level.

### PR DESCRIPTION
**Pull request by duskoo**
Screen Shot 2020-06-08 at 8 57 04 PM

### CODE REVIEW

File: CoEpi/native/NativeCore.swift [hunk 1]
```diff
@@ -43,10 +43,48 @@ extension NativeCore: ServicesBootstrapper {
     // WARN: Must not run long operations. Currently executed synchronously during startup.
     func bootstrap(dbPath: String) -> Result<(), ServicesError> {
         let libResult: LibResult<ArbitraryType>? = bootstrap_core(dbPath)?.toLibResult()
+        NSLog("%@", "blah");
```
Comment (ivnsch) [inline CoEpi/native/NativeCore.swift]: Remove...

File: CoEpi/native/NativeCore.swift [hunk 2]
```diff
@@ -43,10 +43,48 @@ extension NativeCore: ServicesBootstrapper {
...
         return libResult?.toVoidResult().mapErrorToServicesError() ?? libraryFailure()
     }
 }
 
+func logToiOS(string: String){
+    let cs = ConsoleLog.init()
+    cs.d(string, tags: [.db])
+}
+
+func logToiOS(logMessage: CoreLogMessage){
```
Comment (ivnsch) [inline CoEpi/native/NativeCore.swift:54]: Why do we have 2 versions? Are we going to use plain strings or the struct?

File: CoEpi/native/NativeCore.swift [hunk 3]
```diff
@@ -43,10 +43,48 @@ extension NativeCore: ServicesBootstrapper {
...
+                logToiOS(string: cfStr as String)
+            }
+        }
+        NSLog("register_callback returned : %d", x );
         return libResult?.toVoidResult().mapErrorToServicesError() ?? libraryFailure()
     }
 }
 
+func logToiOS(string: String){
+    let cs = ConsoleLog.init()
```
Comment (ivnsch) [inline CoEpi/native/NativeCore.swift]: Use the global `log` (`CompositeLog`: forwards to the `NSLog` and the UI)

File: CoEpi/native/NativeCore.swift [hunk 4]
```diff
@@ -43,10 +43,48 @@ extension NativeCore: ServicesBootstrapper {
...
+    cs.d(string, tags: [.db])
+}
+
+func logToiOS(logMessage: CoreLogMessage){
+    let cs = ConsoleLog.init()
```
Comment (ivnsch) [inline CoEpi/native/NativeCore.swift]: Use the global `log`

File: CoEpi/native/NativeCore.swift [hunk 5]
```diff
@@ -43,10 +43,48 @@ extension NativeCore: ServicesBootstrapper {
...
+    }
+    let text = unmanagedText.takeRetainedValue() as String
+    
+    switch logMessage.level {
+    case 0:
+        cs.v(text, tags: [LogTag.core])
+    case 1:
+        cs.d(text, tags: [LogTag.core])
+    case 2:
+        cs.i(text, tags: [LogTag.db])
```
Comment (ivnsch) [inline CoEpi/native/NativeCore.swift]: Why do we have different tags per level?

(btw, my tag enum probably should be plain strings, to allow core to set tags too. But we can let it like this for 0.3)

File: CoEpiTests/FFISanityTests.swift [hunk 6]
```diff
@@ -37,57 +37,89 @@ class FFISanityTests: XCTestCase {
...
     }
+    
+    func testRegisterCallbackForLogging() {
+        let ret = register_callback { (myInt, myBool, myCFString) in
+            let cfStr: CFString = myCFString!
+            let str = cfStr as String
+            
+            // The first 2 parameters are hardcoded in Rust
+            XCTAssertEqual(myInt, 1)
+            XCTAssertEqual(myBool, true)
```
Comment (ivnsch) [inline CoEpiTests/FFISanityTests.swift]: This shouldn't be part of the logger functionality/tests.

**Review by ivnsch — CHANGES_REQUESTED** [review]

File: CoEpi/native/NativeCore.swift (hunk 2 above)
Comment (duskoo) [reply CoEpi/native/NativeCore.swift:54]: Struct. String were used in first iteration.

Comment (ivnsch) [reply CoEpi/native/NativeCore.swift:54]: Ok, can you please remove the outdated stuff?

File: CoEpiTests/FFISanityTests.swift (hunk 6 above)
Comment (duskoo) [reply CoEpiTests/FFISanityTests.swift]: Method deleted (it was testing string logging)

File: CoEpi/native/NativeCore.swift (hunk 5 above)
Comment (duskoo) [reply CoEpi/native/NativeCore.swift]: Just tryin out if the logic works. 
Ok.

File: CoEpi/native/NativeCore.swift (hunk 2 above)
Comment (duskoo) [reply CoEpi/native/NativeCore.swift:54]: Removed.

File: CoEpi/native/NativeCore.swift (hunk 1 above)
Comment (duskoo) [reply CoEpi/native/NativeCore.swift]: Removed.

File: CoEpi/native/NativeCore.swift (hunk 4 above)
Comment (duskoo) [reply CoEpi/native/NativeCore.swift]: Done.

**Review by ivnsch — APPROVED** [review]

**Pull request merged**